src/wasm/literal.cpp

```diff
@@ -73,40 +73,55 @@ Literal::Literal(const Literal& other) : type(other.type) {
     } else {
       new (&exn) std::unique_ptr<ExceptionPackage>();
     }
-  } else if (other.isGCData()) {
+    return;
+  }
+  if (other.isGCData()) {
     new (&gcData) std::shared_ptr<GCData>(other.gcData);
-  } else if (type.isFunction()) {
+    return;
+  }
+  if (type.isFunction()) {
     func = other.func;
-  } else if (type.isRtt()) {
+    return;
+  }
+  if (type.isRtt()) {
     // Allocate a new RttSupers with a copy of the other's data.
     new (&rttSupers) auto(std::make_unique<RttSupers>(*other.rttSupers));
-  } else {
-    TODO_SINGLE_COMPOUND(type);
-    switch (type.getBasic()) {
-      case Type::i32:
-      case Type::f32:
-      case Type::i31ref:
-        i32 = other.i32;
-        break;
-      case Type::i64:
-      case Type::f64:
-        i64 = other.i64;
-        break;
-      case Type::v128:
-        memcpy(&v128, other.v128, 16);
-        break;
-      case Type::none:
-        break;
-      case Type::externref:
-      case Type::anyref:
-      case Type::eqref:
-        break; // null
-      case Type::funcref:
-      case Type::exnref:
-      case Type::unreachable:
-        WASM_UNREACHABLE("unexpected type");
+    return;
+  }
+  if (type.isRef()) {
+    auto heapType = type.getHeapType();
+    if (heapType.isBasic()) {
+      switch (heapType.getBasic()) {
+        case HeapType::BasicHeapType::any:
+        case HeapType::BasicHeapType::ext:
+        case HeapType::BasicHeapType::eq:
+          return; // null
+        case HeapType::BasicHeapType::i31:
+          i32 = other.i32;
+          return;
+        default:
+          WASM_UNREACHABLE("unexpected heap type");
+      }
     }
   }
+  TODO_SINGLE_COMPOUND(type);
+  switch (type.getBasic()) {
+    case Type::i32:
+    case Type::f32:
+      i32 = other.i32;
+      break;
+    case Type::i64:
+    case Type::f64:
+      i64 = other.i64;
+      break;
+    case Type::v128:
+      memcpy(&v128, other.v128, 16);
+      break;
+    case Type::none:
+      break;
+    default:
+      WASM_UNREACHABLE("unexpected type");
```

Comment:
I generally prefer not having `default` cases for these switches so the compiler can tell help us find all the places to update when we add a new type. Deleting all the BasicTypes for reference types would help make the class of error this PR fixes impossible and would help keep the switches from having too many useless cases. wdyt?

Reply:
You're right, that is better. Done.